lib/SILOptimizer/Mandatory/FlowIsolation.cpp

```diff
@@ -0,0 +1,608 @@
+
+/// Analyzes a function for uses of `self` and records the kinds of isolation
+/// required.
+/// \param selfParam the parameter of \c getFunction() that should be
+/// treated as \c self
+void AnalysisInfo::analyze(const SILArgument *selfParam) {
+  assert(selfParam && "analyzing a function with no self?");
+
+  // Use a worklist to track the uses left to be searched.
+  SmallVector<Operand *, 32> worklist;
+
+  // Seed with direct users of `self`
+  worklist.append(selfParam->use_begin(), selfParam->use_end());
+
+  while (!worklist.empty()) {
+    Operand *operand = worklist.pop_back_val();
+    SILInstruction *user = operand->getUser();
+
+    // First, check if this is an apply that involves `self`
+    if (auto apply = ApplySite::isa(user)) {
+
+      // Check if the callee is a function representing a defer block.
+      if (SILFunction *callee = apply.getCalleeFunction()) {
+        if (auto *dc = callee->getDeclContext()) {
+          if (auto *decl = dyn_cast_or_null<FuncDecl>(dc->getAsDecl())) {
+            if (decl->isDeferBody()) {
+
+              // If we need to analyze the defer first, do so.
+              if (!haveDeferInfo(callee)) {
+                // NOTE: the defer function is not like a method, because it
+                // doesn't satisfy hasSelfParam().
+                auto const* calleeSelfParam =
+                  callee->getArgument(apply.getAppliedArgIndex(*operand));
+
+                // Recursion depth is bounded by the lexical nesting of
+                // defer blocks in the input program.
+                auto &defer = getOrCreateDeferInfo(callee);
+                defer.analyze(calleeSelfParam);
+                defer.solve();
+              }
+
+              auto const& defer = getOrCreateDeferInfo(callee);
+
+              // A defer effectively has one exit block, because it can't throw.
+              // Otherwise, it may never return (e.g., fatalError).
+              // So, we say that this instruction generates nonisolation only
+              // if it can return normally, and if it does, it carries
+              // nonisolation.
+              if (defer.normalReturn) {
+                if (defer.normalReturn->second == State::Nonisolated) {
+                  mark(State::Nonisolated, user);
+                }
+              }
+
+              // If the defer body has any stored property uses, we record that
+              // in the parent by declaring this call-site being a property use.
+              if (defer.hasPropertyUse()) {
+                mark(State::Isolated, user);
+              }
+
+              continue;
+            }
+          }
+        }
+      }
+
+      // For all other call-sites, uses of `self` are nonisolated.
+      mark(State::Nonisolated, user);
+      continue;
+    }
+
+    // Handle non-ApplySite instructions.
+    switch (user->getKind()) {
+      // isolated uses
+      case SILInstructionKind::StoreInst:
+      case SILInstructionKind::LoadInst:
```

Comment:
Formal access should always have an access scope. I'm curious why we need to look for individual load/store instructions. That doesn't see, reliable.

Reply:
Thanks, I hadn't thought of that. I've fixed this and now it uses the access scope, so it can handle other kinds of accesses. I might need to refine this a bit more to hang onto the `ref_element_addr` so that diagnostics underline the right thing. Sometimes it underlines the `=` in an assignment, etc.

Reply:
I had to go with tracking RefElementAddrs, because not all initializers used formal access scopes correctly (I can't recall which ones unfortunately).